api/lib/block-listeners/base-node-subscription.js

```diff
@@ -124,9 +139,92 @@ class BaseNodeSubscription {
       this.store.update({
         validators: validators
       })
+      const storeValidators = Object.values(this.store.validators)
+      if (storeValidators.length > 0) {
+        // now that we have validators in store start polling for validator profiles
+        this.pollForValidatorsProfiles(storeValidators)
+      }
+    })
+  }
+
+  async getValidatorsProfilesFromDB(allValidatorsAddresses) {
+    const allValidatorsProfiles = await this.db.getAllValidatorsProfiles(
+      allValidatorsAddresses,
+      this.network.id
+    )
+    return keyBy(allValidatorsProfiles, `operator_address`)
+  }
+
+  async getAllValidatorsFeed(
+    validators,
+    allValidatorsAddresses,
+    networkList,
+    dataSource
+  ) {
+    const allValidatorsFeed = await this.db.getAccountsNotifications(
+      allValidatorsAddresses,
+      this.network.id
+    )
+    return validators.map((validator) => {
+      const validatorFeed = allValidatorsFeed.filter(
+        ({ resourceId }) => resourceId === validator.operatorAddress
+      )
+      return {
+        ...validator,
+        feed:
+          validatorFeed && Array.isArray(validatorFeed)
+            ? validatorFeed.map((notification) =>
+                dataSource.reducers.notificationReducer(
+                  notification,
+                  networkList
+                )
+              )
+            : []
+      }
     })
   }
 
+  async getAllValidatorsProfiles(validators, dataSource) {
+    const networkList = await db.getNetworks()
+    validators = getRanksForValidators(validators)
+    const allValidatorsAddresses = validators.map(
+      ({ operatorAddress }) => operatorAddress
+    )
+    validators = await this.getAllValidatorsFeed(
+      validators,
+      allValidatorsAddresses,
+      networkList,
+      dataSource
+    )
+    const validatorProfilesDictionary = await this.getValidatorsProfilesFromDB(
+      allValidatorsAddresses
+    )
+    return await Promise.all(
+      validators.map(async (validator) => {
+        const allValidatorDelegations = await dataSource.getAllValidatorDelegations(
+          validator
+        )
+        return dataSource.reducers.validatorProfileReducer(
+          validator,
+          validatorProfilesDictionary[validator.operatorAddress],
+          allValidatorDelegations.length,
+          this.network,
+          validator.feed
```

Comment:
you already pass `validator`, why do you pass this subprop again?

Reply:
true true :+1: